drivers/ethernet/eth_w5500.c

```diff
@@ -433,14 +457,16 @@ static void w5500_set_macaddr(const struct device *dev)
 {
 	struct w5500_runtime *ctx = dev->data;
 
-	/* override vendor bytes */
+#if DT_INST_PROP(0, zephyr_random_mac_address)
 	memset(ctx->mac_addr, '\0', sizeof(ctx->mac_addr));
-	ctx->mac_addr[0] = WIZNET_OUI_B0;
-	ctx->mac_addr[1] = WIZNET_OUI_B1;
-	ctx->mac_addr[2] = WIZNET_OUI_B2;
 	if (ctx->generate_mac) {
 		ctx->generate_mac(ctx->mac_addr);
 	}
+#endif
+	/* override vendor bytes for both random and local_mac_address */
+	ctx->mac_addr[0] = WIZNET_OUI_B0;
+	ctx->mac_addr[1] = WIZNET_OUI_B1;
+	ctx->mac_addr[2] = WIZNET_OUI_B2;
```

Comment:
That looks weird and IMO this changes should be done in a separate commit.

`w5500_random_mac()` passes `WIZNET_OUI_B?` to `gen_random_mac()`, why is it necessary to overwrite it again?
Is it okay to assign only mac_addr[0...2] if there is no `DT_INST_PROP(0, zephyr_random_mac_address)` and why?

Reply:
I touched and tested this code long time ago. If I remember correctly then the reason to override here was that my W5500 hardware chip ignored custom OUI vendor bytes but underlying packet handling and filtering in Zephyr uses provided mac_addr. 

In example if user provided wrong OUI bytes in local_mac_address then the networking didn't work at all.

Reply:
But you are right, mac_addr[0...2] assignment could be moved to #else block if zephyr_random_mac_address is not used at the moment.

The reason I implemented it in that way was that I didn't fully understand why the original code was implemented in a way as it was. In example why to check if ctx->generate_mac is assigned when it is always there and assign the OUI bytes before it, when ctx->generate_mac() call assigns them anyway later. This all rendered the DT provided local_mac_address unusable and random MAC was used always.

The only explanation I could come up was that the original implementer wanted to make ctx->generate_mac function assignable from the application code or somewhere else and for this reason it is safer to override OUI bytes every time.

Reply:
If there is no zephyr_random_mac_address property  and there is a valid MAC address stored in the controller registers, the driver should read MAC address from the controller registers.

Reply:
This is a good idea not to change the MAC if W5500 already has fixed and unique address stored internally. This was my original requirement to get unique and stable MAC and I somewhat remember that I even tried to test it out and it didn't work. I can't be sure at the moment as it all happened 7 months ago. But this should be a separate PR then as my PR is about making possible to change existing MAC from the user space application even if it was assigned in HW or set as random in driver.

Reply:
I tested the custom OUI bytes again and this time there were no issues. No idea, if I did something wrong during testing 7 months ago or had different W5500 HW revision back then when I did the first tests. Reverted my changes to original code but kept the random mac address DT conditional define.

Reply:
> If there is no zephyr_random_mac_address property and there is a valid MAC address stored in the controller registers, the driver should read MAC address from the controller registers.

Checked this from datasheet and after reset the Source Hardware Address (MAC address) is filled with default values 0x00. This means that MAC address has to be updated after every restart from the driver or application.